### PR TITLE
Exercise four - network communication

### PR DESCRIPTION
**Pull request by maz4**

### CODE REVIEW

File: js/network.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+      console.log('Server error: ', message);
+    },
+
+    successHandler: function (message) {
+      console.log('Server responce: ', message);
+    },
+
+    receive: function (onLoad) {
+      var reciveDataUrl = 'https://js.dump.academy/keksobooking/data';
+      var xhrRecive = new XMLHttpRequest();
```
Comment (mibu5) [inline js/network.js]: xhrRecive -> xhr

File: js/network.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+      xhrRecive.addEventListener('timeout', xhrTimeoutHandler);
+
+      xhrRecive.timeout = 10000;
+      xhrRecive.open('GET', reciveDataUrl);
+      xhrRecive.send();
+    },
+
+    send: function (data) {
+      var sendUrl = 'https://js.dump.academy/keksobooking';
+      var xhrSend = new XMLHttpRequest();
```
Comment (mibu5) [inline js/network.js]: xhrSend -> xhr

File: js/network.js [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  return {
+    errorHandler: function (message) {
+      console.log('Server error: ', message);
+    },
+
+    successHandler: function (message) {
+      console.log('Server responce: ', message);
+    },
+
+    receive: function (onLoad) {
```
Comment (mibu5) [inline js/network.js]: onError handler should be passed to this function as an argument.

File: js/network.js [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+      xhrRecive.addEventListener('load', xhrReciveHandler);
+      xhrRecive.addEventListener('error', xhrErrorHandler);
+      xhrRecive.addEventListener('timeout', xhrTimeoutHandler);
+
+      xhrRecive.timeout = 10000;
+      xhrRecive.open('GET', reciveDataUrl);
+      xhrRecive.send();
+    },
+
+    send: function (data) {
```
Comment (mibu5) [inline js/network.js]: should accept two more arguments onSuccess handler and onError handler

File: js/network.js [hunk 5]
```diff
@@ -0,0 +1,69 @@
+'use strict';
+window.network = (function () {
+
+  return {
+    errorHandler: function (message) {
+      console.log('Server error: ', message);
+    },
+
+    successHandler: function (message) {
+      console.log('Server responce: ', message);
```
Comment (mibu5) [inline js/network.js]: And use templates to show messages to the user. Don't need console.logs

File: js/network.js (hunk 1 above)
Comment (maz4) [reply js/network.js]: done

File: js/network.js (hunk 2 above)
Comment (maz4) [reply js/network.js]: done

File: js/network.js (hunk 3 above)
Comment (maz4) [reply js/network.js]: done

File: js/network.js (hunk 4 above)
Comment (maz4) [reply js/network.js]: done

File: js/network.js (hunk 5 above)
Comment (maz4) [reply js/network.js]: done

File: js/form.js [hunk 6]
```diff
@@ -2,22 +2,25 @@
...
   var propertyTypeElement = adFormContainer.querySelector('#type');
   var checkinElement = adFormContainer.querySelector('#timein');
   var checkoutElement = adFormContainer.querySelector('#timeout');
   var capacityElement = adFormContainer.querySelector('#capacity');
   var roomsElement = adFormContainer.querySelector('#room_number');
+  var submitButton = adFormContainer.querySelector('.ad-form__submit');
+  var errorTemplate = document.querySelector('#error').content.querySelector('.error');
+  var successTemplate = document.querySelector('#success').content.querySelector('.success');
+
+  var successMessage = null;
```
Comment (mibu5) [inline js/form.js]: do we really need a global message?

File: js/form.js [hunk 7]
```diff
@@ -65,15 +68,81 @@ window.form = (function () {
...
+  function escButtonHandler(event) {
+    if (event.keyCode === 27) {
+      successMessage.classList.add('hidden');
+      event.target.removeEventListener('keydown', escButtonHandler);
+      document.removeEventListener('click', hideSuccessHandler);
+    }
+  }
+
+  function successSendHandler() {
+    if (!successMessage) {
```
Comment (mibu5) [inline js/form.js]: you need to create it every time, and destroy every time after you close it. Same as you're doing with the Card.

File: js/form.js [hunk 8]
```diff
@@ -65,15 +68,81 @@ window.form = (function () {
...
+    function errorButtonHandler() {
+      var data = new FormData(adFormContainer);
+      window.network.send(data, successSendHandler, errorSendHandler);
+
+      errorButton.removeEventListener('click', errorButtonHandler);
+      errorMessageElement.parentElement.removeChild(errorMessageElement);
+    }
+    errorButton.addEventListener('click', errorButtonHandler);
+
+    setTimeout(function () {
```
Comment (mibu5) [inline js/form.js]: Was it part of the requirements?

File: js/map.js [hunk 9]
```diff
@@ -137,4 +194,33 @@ window.map = (function () {
 
   mapPinsMainElement.addEventListener('mousedown', mouseDownHandler);
   mapElement.addEventListener('mouseup', pinsHandler);
+
+  return {
+    resetMap: function () {
```
Comment (mibu5) [inline js/map.js]: Let have it as a separate function, and return it here as:
{
resetMap: resetMap
}

File: js/map.js [hunk 10]
```diff
@@ -137,4 +194,33 @@ window.map = (function () {
...
+      mapPinsMainElement.style.left = MAP_PIN_LEFT + 'px';
+      mapElement.classList.add('map--faded');
+      adFormContainer.classList.add('ad-form--disabled');
+
+      if (mapElement.querySelector('.popup')) {
+        removeCard();
+      }
+
+      [].forEach.call(mapPins, function (elem) {
+        if (elem.className === 'map__pin map__pin--main') {
```
Comment (mibu5) [inline js/map.js]: You can select just map__pin:not(.map__pin--main) in the first place.
Also what do you think about saving the pins to an array when you create them and then just iterate over the array here?

File: js/network.js [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+          onError(event.target.status);
+        }
+      }
+
+      function xhrErrorHandler(event) {
+        onError(event.target.status);
+      }
+
+      function xhrTimeoutHandler() {
+        console.log('Connection time out');
```
Comment (mibu5) [inline js/network.js]: should pass an error to the onError callBack

File: js/network.js [hunk 12]
```diff
@@ -0,0 +1,60 @@
+'use strict';
+window.network = (function () {
+  return {
+    receive: function (onLoad, onError) {
+      var reciveDataUrl = 'https://js.dump.academy/keksobooking/data';
```
Comment (mibu5) [inline js/network.js]: typo

File: js/network.js [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+      var xhr = new XMLHttpRequest();
+
+      function xhrReciveHandler(event) {
```
Comment (mibu5) [inline js/network.js]: typo

File: js/network.js [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+
+    send: function (data, onSuccess, onError) {
+      var sendUrl = 'https://js.dump.academy/keksobooking';
+      var xhr = new XMLHttpRequest();
+      xhr.open('POST', sendUrl);
+
+      function xhrSendHandler() {
+        if (xhr.status >= 200 && xhr.status < 400) {
+          onSuccess();
+          document.querySelector('.ad-form').reset();
```
Comment (mibu5) [inline js/network.js]: network module should not know anything about the form. should be done in the callback

File: js/network.js [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+          window.form.submitButton.disabled = false;
```
Comment (mibu5) [inline js/network.js]: same as above. does not belong here

File: js/network.js [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+
+        } else {
+          onError();
+          window.form.submitButton.disabled = false;
```
Comment (mibu5) [inline js/network.js]: does not belong here

File: js/network.js [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+        }
+      }
+
+      function xhrErrorHandler() {
+        onError();
+        window.form.submitButton.disabled = false;
```
Comment (mibu5) [inline js/network.js]: does not belong here

File: js/network.js [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+      xhr.open('POST', sendUrl);
+
+      function xhrSendHandler() {
+        if (xhr.status >= 200 && xhr.status < 400) {
+          onSuccess();
+          document.querySelector('.ad-form').reset();
+          window.form.submitButton.disabled = false;
+
+        } else {
+          onError();
```
Comment (mibu5) [inline js/network.js]: should we pass an error here?

File: js/network.js [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+          window.form.submitButton.disabled = false;
+        }
+      }
+
+      function xhrErrorHandler() {
+        onError();
```
Comment (mibu5) [inline js/network.js]: should we pass an error so we know what happened?

**Review by mibu5 — CHANGES_REQUESTED** [review]

File: js/form.js (hunk 6 above)
Comment (maz4) [reply js/form.js]: done

File: js/form.js (hunk 7 above)
Comment (maz4) [reply js/form.js]: done

File: js/form.js (hunk 8 above)
Comment (maz4) [reply js/form.js]: yes

File: js/map.js (hunk 9 above)
Comment (maz4) [reply js/map.js]: done

File: js/map.js (hunk 10 above)
Comment (maz4) [reply js/map.js]: done

File: js/network.js (hunk 12 above)
Comment (maz4) [reply js/network.js]: done

File: js/network.js (hunk 13 above)
Comment (maz4) [reply js/network.js]: done

File: js/network.js (hunk 11 above)
Comment (maz4) [reply js/network.js]: done

File: js/network.js (hunk 18 above)
Comment (maz4) [reply js/network.js]: done

File: js/network.js (hunk 19 above)
Comment (maz4) [reply js/network.js]: done

File: js/network.js (hunk 14 above)
Comment (maz4) [reply js/network.js]: done

File: js/network.js (hunk 15 above)
Comment (maz4) [reply js/network.js]: done

File: js/network.js (hunk 16 above)
Comment (maz4) [reply js/network.js]: done

File: js/network.js (hunk 17 above)
Comment (maz4) [reply js/network.js]: done

File: js/form.js [hunk 20]
```diff
@@ -65,11 +66,83 @@ window.form = (function () {
...
+    document.querySelector('body').appendChild(successTemplate);
+    var successMessageElement = document.querySelector('.success');
+    successMessageElement.addEventListener('click', successMsgHandler);
+    document.addEventListener('keydown', escButtonHandler);
+    submitButton.disabled = false;
+    resetPage();
+  }
+
+  function errorSendHandler(errorText) {
+    document.querySelector('body').appendChild(errorTemplate);
```
Comment (mibu5) [inline js/form.js]: clone template here using cloneNode. Treat you template as an example.

File: js/form.js [hunk 21]
```diff
@@ -65,11 +66,83 @@ window.form = (function () {
...
+    var errorButton = document.querySelector('.error__button');
```
Comment (mibu5) [inline js/form.js]: after you've cloned and created a new template, use the new element as the context to elements within

`var errorButton = errorMessageElement.querySelector('.error__button');`

File: js/form.js [hunk 22]
```diff
@@ -65,11 +66,83 @@ window.form = (function () {
...
+    var errorMessageElement = document.querySelector('.error');
```
Comment (mibu5) [inline js/form.js]: won't need this as it will be the copied template.

File: js/form.js [hunk 23]
```diff
@@ -65,11 +66,83 @@ window.form = (function () {
...
+    var errorMessageText = document.querySelector('.error__message');
```
Comment (mibu5) [inline js/form.js]: errorMessageElement.querySelector('.error__message')

File: js/form.js [hunk 24]
```diff
@@ -65,11 +66,83 @@ window.form = (function () {
...
+
+    function errorButtonHandler() {
+      var data = new FormData(adFormContainer);
+      window.network.send(data, successSendHandler, errorSendHandler);
+
+      errorButton.removeEventListener('click', errorButtonHandler);
+      errorMessageElement.parentElement.removeChild(errorMessageElement);
+    }
+
+    errorMessageText.innerHTML = 'Sending error: ' + errorText;
```
Comment (mibu5) [inline js/form.js]: use TextContent if it's just a text.

File: js/form.js [hunk 25]
```diff
@@ -65,11 +66,83 @@ window.form = (function () {
...
+  function escButtonHandler(event) {
+    if (event.keyCode === 27) {
+      document.querySelector('.success').removeEventListener('click', successMsgHandler);
+      document.removeEventListener('keydown', escButtonHandler);
+      removeMessage();
+    }
+  }
+
+  function successSendHandler() {
+    document.querySelector('body').appendChild(successTemplate);
```
Comment (mibu5) [inline js/form.js]: similar as with errorSendHandler bellow.

File: js/map.js [hunk 26]
```diff
@@ -13,19 +15,73 @@ window.map = (function () {
...
+    receivedData[7].offer.title = 'Apartment in the center';
+    receivedData[7].offer.description = 'It is beautiful, light and cozy. It can accommodate 5 people. Coffee and cookies provided for free.';
+    receivedData[8].offer.title = 'Quiet apartment near the subway';
+    receivedData[8].offer.description = 'Apartment on the first floor. Neighbors are quiet. For anyone who can not stand the hustle and bustle.';
+    receivedData[9].offer.title = 'Cute nest for Anime fans';
+    receivedData[9].offer.description = 'Enjoy your time nad please do not disturb others.';
+    return receivedData;
+  }
+
+  function errorReceiveHandler(errorText) {
```
Comment (mibu5) [inline js/map.js]: Looks like you're using the same code as in form module. 
Let's create a separate module to handle messages and reuse it across.

**Review by mibu5 — CHANGES_REQUESTED** [review]

File: js/form.js (hunk 20 above)
Comment (maz4) [reply js/form.js]: done

File: js/form.js (hunk 21 above)
Comment (maz4) [reply js/form.js]: done

File: js/form.js (hunk 22 above)
Comment (maz4) [reply js/form.js]: done

File: js/form.js (hunk 23 above)
Comment (maz4) [reply js/form.js]: done

File: js/form.js (hunk 24 above)
Comment (maz4) [reply js/form.js]: done

File: js/form.js (hunk 25 above)
Comment (maz4) [reply js/form.js]: done

File: js/map.js (hunk 26 above)
Comment (maz4) [reply js/map.js]: done

File: js/error.js [hunk 27]
```diff
@@ -0,0 +1,24 @@
...
+  var errorMessageElement = errorTemplate.cloneNode(true);
+  var errorButton = errorMessageElement.querySelector('.error__button');
+  var errorText = errorMessageElement.querySelector('.error__message');
+
+  return {
+    errorHandler: function (error, errBtnHandler) {
+      errorText.innerText = 'Sending error: ' + error;
+      errorButton.addEventListener('click', errBtnHandler);
+      document.querySelector('body').appendChild(errorMessageElement);
+      var errorElement = document.querySelector('body').querySelector('.error');
```
Comment (mibu5) [inline js/error.js]: errorElement is the same as errorMessageElement

File: js/error.js [hunk 28]
```diff
@@ -0,0 +1,24 @@
+'use strict';
+window.error = (function () {
+  var errorTemplate = document.querySelector('#error').content.querySelector('.error');
+  var errorMessageElement = errorTemplate.cloneNode(true);
```
Comment (mibu5) [inline js/error.js]: line 4-6 should be inside errorHandler function. You create new template every time you generate an error screen

File: js/form.js [hunk 29]
```diff
@@ -2,22 +2,23 @@
...
-
   var formAddressInput = adFormContainer.querySelector('#address');
   var priceElement = adFormContainer.querySelector('#price');
   var propertyTypeElement = adFormContainer.querySelector('#type');
   var checkinElement = adFormContainer.querySelector('#timein');
   var checkoutElement = adFormContainer.querySelector('#timeout');
   var capacityElement = adFormContainer.querySelector('#capacity');
   var roomsElement = adFormContainer.querySelector('#room_number');
+  var submitButton = adFormContainer.querySelector('.ad-form__submit');
+  var errorTemplate = document.querySelector('#error').content.querySelector('.error');
```
Comment (mibu5) [inline js/form.js]: not used

File: js/map.js [hunk 30]
```diff
@@ -13,19 +15,59 @@ window.map = (function () {
     x: 0,
     y: 0
   };
+  var isPageActive = false;
+  var data = [];
+  var errorTemplate = document.querySelector('#error').content.querySelector('.error');
```
Comment (mibu5) [inline js/map.js]: not used

File: js/error.js [hunk 31]
```diff
@@ -0,0 +1,24 @@
+'use strict';
+window.error = (function () {
```
Comment (mibu5) [inline js/error.js]: the Idea is that it's not a error module. 
It's a message handling module that adds and removes message. 
I think you need to add a success message here as well that handles successTemplate creation.

**Review by mibu5 — CHANGES_REQUESTED** [review]

File: js/error.js (hunk 27 above)
Comment (maz4) [reply js/error.js]: not the same as errorMessageElement is inside the variable which is not added to the body yet,
where errorElement is selected from the body after adding errorMessageElement to the DOM

File: js/error.js (hunk 28 above)
Comment (maz4) [reply js/error.js]: done

File: js/error.js (hunk 27 above)
Comment (mibu5) [reply js/error.js]: You're querying the same element after you've appended it to the body. It is not optimal. Don't see how it is different. 
line 12 `document.querySelector('body').appendChild(errorMessageElement);`
line 12 `document.querySelector('body').querySelector('.error');`

File: js/error.js [hunk 32]
```diff
@@ -0,0 +1,24 @@
...
+
+  return {
+    errorHandler: function (error, errBtnHandler) {
+      errorText.innerText = 'Sending error: ' + error;
+      errorButton.addEventListener('click', errBtnHandler);
+      document.querySelector('body').appendChild(errorMessageElement);
+      var errorElement = document.querySelector('body').querySelector('.error');
+
+      setTimeout(function () {
+        var errorButtonElement = document.querySelector('.error__button');
```
Comment (mibu5) [inline js/error.js]: using document is not optimal here neither use `errorMessageElement.querySelector('.error__button')`

File: js/error.js [hunk 33]
```diff
@@ -0,0 +1,24 @@
...
+        if (errorButtonElement) {
+          errorButtonElement.removeEventListener('click', errBtnHandler);
+          document.querySelector('body').removeChild(errorElement);
```
Comment (mibu5) [inline js/error.js]: `errorMessageElement.parentNode.removeChild(errorMessageElement)`

File: js/error.js [hunk 34]
```diff
@@ -0,0 +1,24 @@
...
+  var errorTemplate = document.querySelector('#error').content.querySelector('.error');
+  var errorMessageElement = errorTemplate.cloneNode(true);
+  var errorButton = errorMessageElement.querySelector('.error__button');
+  var errorText = errorMessageElement.querySelector('.error__message');
+
+  return {
+    errorHandler: function (error, errBtnHandler) {
+      errorText.innerText = 'Sending error: ' + error;
+      errorButton.addEventListener('click', errBtnHandler);
+      document.querySelector('body').appendChild(errorMessageElement);
```
Comment (mibu5) [inline js/error.js]: don't query body every time, cache it in a variable outside the function
`errorElementContainer = document.querySelector('body')`
and then
`errorElementContainer.appendChild(errorMessageElement)`

File: js/map.js [hunk 35]
```diff
@@ -81,9 +123,14 @@ window.map = (function () {
     document.addEventListener('mousemove', mouseMoveHandler);
     document.addEventListener('mouseup', mouseUpHandler);
 
-    mapElement.classList.remove('map--faded');
-    adFormContainer.classList.remove('ad-form--disabled');
-    window.form.enableFormInputs();
+    if (!isPageActive) {
+      isPageActive = true;
+      // window.network.receive(receiveDataHandler, errorReceiveHandler);
```
Comment (mibu5) [inline js/map.js]: comment here

File: js/network.js [hunk 36]
```diff
@@ -0,0 +1,56 @@
...
+
+      function xhrTimeoutHandler(event) {
+        onError(event.target.status, errBtnHandler);
+      }
+
+      xhr.addEventListener('load', xhrReceiveHandler);
+      xhr.addEventListener('error', xhrErrorHandler);
+      xhr.addEventListener('timeout', xhrTimeoutHandler);
+
+      xhr.timeout = 10000;
```
Comment (mibu5) [inline js/network.js]: Why did you decide to use such a long timeout? 10 seconds?

File: js/form.js (hunk 29 above)
Comment (maz4) [reply js/form.js]: removed

File: js/map.js (hunk 30 above)
Comment (maz4) [reply js/map.js]: removed

File: js/error.js (hunk 31 above)
Comment (maz4) [reply js/error.js]: done

File: js/error.js (hunk 32 above)
Comment (maz4) [reply js/error.js]: done

File: js/error.js (hunk 33 above)
Comment (maz4) [reply js/error.js]: done

File: js/error.js (hunk 34 above)
Comment (maz4) [reply js/error.js]: done

File: js/map.js (hunk 35 above)
Comment (maz4) [reply js/map.js]: done

File: js/network.js (hunk 36 above)
Comment (maz4) [reply js/network.js]: done

**Review by mibu5 — APPROVED** [review]

**Pull request merged**